### PR TITLE
feat(rest) : Rest end point for project ECC Export Spreadsheet

### LINKED ISSUE #2462: feat(Rest) : Rest end point for project ECC Export Spreadsheet
### Description

Write an Rest API for project page ECC Export Spreadsheet

#### Screenshots

image

### PR DESCRIPTION
**Pull request by keerthi-bl**
### Description

 Rest end point for project ECC Export Spreadsheet

#### How to reproduce

Closes #2462 

#### Screenshots ( if applicable )

Add screenshots to help explain your problem.

### Versions

* Docker version OR
* Last commit id on main

#### SW360 logs

* With docker through `docker logs sw360`
* From bare metal / vm install system under /var/log/sw360/sw360.log and /var/log/tomcat/error.log

### CODE REVIEW

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/report/SW360ReportController.java [hunk 1]
```diff
@@ -76,15 +67,15 @@ public RepositoryLinksResource process(RepositoryLinksResource resource) {
...
+    @GetMapping(value = SW360Constants.REPORTS_URL)
     public void getProjectReport(
             @Parameter(description = "Projects with linked releases.")
             @RequestParam(value = "withlinkedreleases", required = false, defaultValue = "false") boolean withLinkedReleases,
             @Parameter(description = "Report download format.", schema = @Schema(allowableValues = {"xls", "xlsx"}))
             @RequestParam(value = "mimetype", required = false, defaultValue = "xlsx") String mimeType,
             @Parameter(description = "Project id.")
             @RequestParam(value = "projectId", required = false) String projectId,
-            @Parameter(description = "Module name.", schema = @Schema(allowableValues = {PROJECTS, COMPONENTS}))
+            @Parameter(description = "Module name.", schema = @Schema(allowableValues = {SW360Constants.PROJECTS, SW360Constants.COMPONENTS}))
```
Comment (GMishx) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/report/SW360ReportController.java:78]: Please update the `allowableValues` to also include `LICENSES` and `PROJECT_RELEASE_SPREADSHEET_WITH_ECCINFO`

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/report/SW360ReportController.java [hunk 2]
```diff
@@ -207,15 +202,15 @@ private boolean validateMimeType(String mimeType) {
...
                             schema = @Schema(type = "string", format = "binary"))
             )}
     )
-    @GetMapping(value = REPORTS_URL + "/download")
+    @GetMapping(value = SW360Constants.REPORTS_URL + "/download")
     public void downloadExcel(
             HttpServletRequest request,
             HttpServletResponse response,
-            @Parameter(description = "Module name.", schema = @Schema(allowableValues = {PROJECTS, COMPONENTS}))
+            @Parameter(description = "Module name.", schema = @Schema(allowableValues = {SW360Constants.PROJECTS, SW360Constants.COMPONENTS}))
```
Comment (GMishx) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/report/SW360ReportController.java:213]: Please update `allowableValues` to also include `LICENSES`.

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/report/SW360ReportService.java [hunk 3]
```diff
@@ -33,7 +44,7 @@ public class SW360ReportService {
     LicenseService.Iface licenseClient = thriftClients.makeLicenseClient();
 
     public ByteBuffer getProjectBuffer(User user, boolean extendedByReleases, String projectId) throws TException {
-        if (projectId != null && validateProject(projectId, user)) {
+        if (projectId != null && !validateProject(projectId, user)) {
```
Comment (GMishx) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/report/SW360ReportService.java:47]: Shouldn't it be "If project is null or it is invalid"?
```suggestion
        if (projectId != null || !validateProject(projectId, user)) {
```

File: rest/resource-server/src/test/java/org/eclipse/sw360/rest/resourceserver/restdocs/ProjectSpecTest.java [hunk 4]
```diff
@@ -2583,4 +2584,22 @@ public void should_document_update_project_with_network() throws Exception {
...
+                        .header("Authorization", "Bearer " + accessToken)
+                        .param("mimetype", "xlsx")
+                        .param("module", SW360Constants.PROJECT_RELEASE_SPREADSHEET_WITH_ECCINFO)
+                        .param("projectId", project.getId())
+                        .accept(MediaTypes.HAL_JSON))
+                .andExpect(status().isOk())
+                .andDo(this.documentationHandler.document(
+                        requestParameters(
+                                parameterWithName("mimetype").description("Projects download format. Possible values are `<xls|xlsx>`"),
+                                parameterWithName("module").description("module represent the project or component. Possible values are `<components|projects>`"),
```
Comment (GMishx) [inline rest/resource-server/src/test/java/org/eclipse/sw360/rest/resourceserver/restdocs/ProjectSpecTest.java:2601]: Possible value should include `projectReleaseSpreadSheetWithEcc` as it is passed in the sample request.

**Review by GMishx — CHANGES_REQUESTED** [review]
Please rebase with latest main as well.

**Comment by GMishx** [comment]
Any updates @keerthi-bl ?

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/report/SW360ReportService.java (hunk 3 above)
Comment (keerthi-bl) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/report/SW360ReportService.java:118]: If projectId==null, No need to check whether its a valid project or not. If its not a null and then check its a valid project or not, based on that next logic will run.

Comment (GMishx) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/report/SW360ReportService.java:118]: @keerthi-bl Ah, but with current logic, if the `projectId == null` then the first condition becomes false and will not check for `validateProject()` and short-circuits the `if` to false. Thus, no exception will be thrown.

If the condition is rewritten as `if (projectId == null || !validateProject(projectId, user))`, the even if the `projectId == null` or if it is not a valid project, the exception is thrown.

If the exception has to be thrown when the `projectId == null`, then the `if` condition above has to be rewritten.

Comment (keerthi-bl) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/report/SW360ReportService.java:118]: @GMishx :

The method is calling from 2 different places. One to fetch project based on project id and one with based on user. If the method called for user, i no need to consider it for a valid project or not, thus projectId will be null based on that user condition.

Comment (keerthi-bl) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/report/SW360ReportService.java:122]: @GMishx 

As per the discussion, comments added. Let me know, can we mark it as resolved.

**Review by GMishx — APPROVED** [review]
Changes looks good.

**Review by GMishx — APPROVED** [review]
Tested, working as expected.

**Pull request merged**